### ISSUE TITLE
Is there any support for WellKnownTypes.Value?

### ISSUE THREAD

**Issue by erikljung** [protobuf-net/protobuf-net.Grpc#291]
When using _SchemaGenerator(type)_ to generate a .proto where the _type_ contains a _WellKnownTypes.Value_, the only output I get in the .proto is `message Value { }`. Is there any support for this, or is there some way to implement support for this on my own?

**Comment by mgravell** [comment]
Where did `WellKnownTypes.Value` come from? I might be misremembering, but I don't *think* protobuf-net has specific support for that currently - largely a function of available time vs demand. Are you perhaps using the Google.Protobuf version? if so: that won't work with protobuf-net

**Comment by erikljung** [comment]
Yes it's from Google.Protobuf.WellKnownTypes. Do you have a suggestion on what to do instead? I would like to represent any value ("object").

**Comment by mgravell** [comment]
Right; so *that* wouldn't work directly even with Google.Protobuf - you'd need to use reflection or similar to unroll your arbitrary `object` into the `Struct` API.

Implementing custom `Struct` support in protobuf-net isn't technically *hard* - but it would require some thought and testing; but hypothetically, we could:

- implement `Struct`+`Value` support
- add custom wrap/unwrap support that converts a type to/from `Struct`, presumably recursively

i.e.

```
Foo obj = ... // 
Struct val = Struct.From<T>(obj); // map all public instance fields/properties by name
// not shown: use `val` in some API, perhaps via gRPC
// ...
// then later:
Foo typed = val.To<T>(); // map back into a new instance of T
```

However, I should emphasize that `Struct`/`Value` by definition loses a lot of the advantages of protobuf, and will take a lot more data on the wire, and be less performant.

---

Might I suggest an alternative? protobuf-net has support for a: discriminated unions (for example, of a set of known types), and b: inheritance. Both of these can be used (separately) to achieve most scenarios where the *specific* type isn't known, but where it is known to be one of a few *expected* types.

**Comment by erikljung** [comment]
You know what, I realize that I already know which Type the incoming value should be (converted to). So I could either go with your idea with oneof (right?), or I can simply try to parse it from a string, handle it with JSON or create a custom message type. In my case I need to parse strings to GUID anyway, so I'm already needing special handling for that. So let it be, and thanks for your quick and thorough response.
